src/stylesheet-registry-context.js

```diff
@@ -0,0 +1,35 @@
+import React from 'react'
+
+const defaultInstance = {}
+
+const methods = [
+  'add',
+  'remove',
+  'update',
+  'flush',
+  'cssRules',
+  'createComputeId',
+  'createComputeSelector',
+  'getIdAndRules',
+  'selectFromServer'
+]
+
+methods.forEach(methodName => {
+  defaultInstance[methodName] = () => {
+    throw new Error(
+      `StyleSheetRegistry.${methodName}: The style registry can no longer be referenced statically. You must wrap each render with a StyleSheetRegistryContext.Provider, and provide a unique registry for each render`
+    )
+  }
+})
+
+const StyleSheetRegistryContext = React.createContext(defaultInstance)
+
+export default StyleSheetRegistryContext
+
+export function wrapElementWithProvider(element, registry) {
```

Comment:
I think that this is unnecessary as users can do 

```jsx
<StyleSheetRegistryContext.Provider value={registry}>
  {children}
</StyleSheetRegistryContext.Provider>
```

Reply:
In fact here is what I would do in `style.js`

```jsx
import StyleSheetRegistry from './stylesheet-registry'

const ctx = React.createContext()

export default class JSXStyle extends Component {
  // ...
}

export function Provider({ registry, children }) {
  return (
    <ctx.Provider value={registry || new StyleSheetRegistry()}>
      {children}
    </ctx.Provider>
  )
}
```

Reply:
I think you should move away from the singleton magic. It's just going to get more and more error prone as React continues to move towards async rendering / streaming rendering on the server.

Mutable singletons are bad. It shouldn't be your default.

Reply:
If people need to do async rendering / streaming would still use multiple providers with their own instances of the registry. The ones who don't would need a default that is exactly the singleton. We can do that for them here and make their lives easier.